modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/XmlProcessor.java

```diff
@@ -0,0 +1,942 @@
+     *  <li>Other node types: returns text content as fallback
+     * </ul>
+     *
+     * @param node the DOM node to extract text from
+     * @return the text content of the node, or null if node is null
+     */
+    private String getNodeValue(Node node) {
+        if (node == null) {
+            return null;
+        }
+
+        return switch (node.getNodeType()) {
+            case Node.ATTRIBUTE_NODE, Node.CDATA_SECTION_NODE, Node.TEXT_NODE -> node.getNodeValue();
+            default -> node.getTextContent();
+        };
+    }
+
+    /**
+     * Applies force_array logic to ensure all fields are arrays when enabled.
+     *
+     * Behavior:<ul>
+     *  <li>If force_array is false: returns content unchanged
+     *  <li>If force_array is true and content is already a List: returns content unchanged
+     *  <li>If force_array is true and content is not a List: wraps content in a new ArrayList
+     *  <li>Handles null content appropriately (wraps null in array if force_array is true)
+     * </ul>
+     *
+     * @param elementName the name of the element (for context, not used in current implementation)
+     * @param content the content to potentially wrap in an array
+     * @return the content, optionally wrapped in an array based on force_array setting
+     */
+    private Object applyForceArray(String elementName, Object content) {
+        if (forceArray && (content instanceof List) == false) {
+            List<Object> arrayContent = new ArrayList<>();
+            arrayContent.add(content);  // Add content even if it's null (for empty elements)
+            return arrayContent;
+        }
+        return content;
+    }
+
+    /**
+     * Evaluates precompiled XPath expressions against a DOM document and adds results to the ingest document.
+     *
+     * Features:<ul>
+     *  <li>Uses precompiled XPath expressions for optimal performance
+     *  <li>Extracts text values from matched nodes (elements, attributes, text nodes)
+     *  <li>Single matches stored as strings, multiple matches as string arrays
+     * </ul>
+     *
+     * @param ingestDocument the ingest document to add XPath results to
+     * @param xmlDocument the DOM document to evaluate XPath expressions against
+     * @throws XPathExpressionException if XPath processing fails
+     */
+    private void processXPathExpressionsFromDom(IngestDocument ingestDocument, Document xmlDocument) throws XPathExpressionException {
+        // Use precompiled XPath expressions for optimal performance
+        for (Map.Entry<String, XPathExpression> entry : compiledXPathExpressions.entrySet()) {
+            String targetFieldName = entry.getKey();
+            XPathExpression compiledExpression = entry.getValue();
+            Object result = compiledExpression.evaluate(xmlDocument, XPathConstants.NODESET);
+
+            if (result instanceof NodeList nodeList) {
+                List<String> values = new ArrayList<>();
```

Comment:
It might be clearer if you split the code into the 0-1 case (where nodeList.getLength() <= 1), and where you do not have to allocate an `ArrayList` at all, and the 0-N case (where nodeList.getLength() > 1), and where you do need to allocate an `ArrayList` (but _might_ still end up only having a single result).

This is partially a code clarity point and partially a performance point. (I haven't benchmarked this, but avoiding  potentially many un-necessary ArrayList allocations per document is almost certainly a win at the margins.)

Reply:
I think @samxbr handled this comment via 998cae6b4ba7dfc0246f947118a17dbabedbff1d.